internal/bsr/bsr.go

```diff
@@ -237,6 +238,173 @@ func OpenSession(ctx context.Context, sessionRecordingId string, f storage.FS, k
 	return session, nil
 }
 
+type Validation struct {
+	Name               string
+	Type               ContainerType
+	Error              error
+	ChecksumValidation ChecksumValidation
+	SubContainer       []Validation `json:"sub_container,omitempty"`
+}
+
+type ValidationSummary struct {
+	SessionRecordingId string
+	Valid              bool
+	FailedChecksums    map[ContainerType]map[string]ChecksumValidation
+}
+
+// Validate retrieves a BSR from storage using the sessionRecordingId and validates the BSR.
+// All files and sub container files will be verified by comparing it against checksums for
+// each file in SHA256SUM file.
+//
+// Validation will continue even if there's an error encountered during validation.
+// The validation error will be added to the Validation struct Error field.
+func Validate(ctx context.Context, sessionRecordingId string, f storage.FS, keyUnwrapFn kms.KeyUnwrapCallbackFunc) (Validation, ValidationSummary, error) {
+	const op = "bsr.Validate"
+	validation := Validation{
+		Name:         sessionRecordingId,
+		Type:         SessionContainer,
+		SubContainer: []Validation{},
+	}
+
+	validationSummary := ValidationSummary{
+		SessionRecordingId: sessionRecordingId,
+		Valid:              true,
+	}
+
+	switch {
+	case sessionRecordingId == "":
+		validationSummary.Valid = false
+		return validation, validationSummary, fmt.Errorf("%s: missing session recording id: %w", op, ErrInvalidParameter)
+	case f == nil:
+		validationSummary.Valid = false
+		return validation, validationSummary, fmt.Errorf("%s: missing storage: %w", op, ErrInvalidParameter)
+	case keyUnwrapFn == nil:
+		validationSummary.Valid = false
+		return validation, validationSummary, fmt.Errorf("%s: missing key unwrap function: %w", op, ErrInvalidParameter)
+	}
+
+	session, err := OpenSession(ctx, sessionRecordingId, f, keyUnwrapFn)
+	if err != nil {
+		validationSummary.Valid = false
+		validationError := fmt.Errorf("%s: failed to retrieve session for %s: %w", op, sessionRecordingId, err)
+		validation.Error = errors.Join(validation.Error, validationError)
+		return validation, validationSummary, validationError
+	}
+
+	// Validate session
+	sessionChecksumValidation, err := session.container.ValidateChecksums(ctx)
+	if err != nil {
+		validationSummary.Valid = false
+		validation.Error = errors.Join(validation.Error, fmt.Errorf("%s: failed to validate session for %s: %w", op, sessionRecordingId, err))
+	}
+
+	validationSummary = updateValidationSummary(ctx, SessionContainer, sessionRecordingId, sessionChecksumValidation, validationSummary)
+
+	validation.ChecksumValidation = sessionChecksumValidation
+
+	conns := session.Meta.connections
+
+	for connId := range conns {
+		// Get connection id
+		connKey := connId
+		lastDotIndex := strings.LastIndex(connId, ".")
```

Comment:
```suggestion
		lastDotIndex := strings.LastIndex(connId, ".connection")
```
Might make this clearer to others in the future